feat: sample app architecture

IvayloGrahlyov opened the pull request.

EmilIvanov commented:
```
app doesn't start
  - missing modules
  - path aliases don't work

unused directories
  - delete

index
  - keep bootstrap function

app
  - db layer ?!
  - middleware - remove factory

auth
  - separate module
  - check express-jwt and passport-jwt

common
  - remove/rename/split

services folder

api versioning
  - prefix routes

api/users/v1/
  - remove controller
  - add file per route

add request schema to route

remove api/sample-module
```

IvayloGrahlyov commented:
In last commit:
- deleted most of empty folders (I think only the __data__) ones are left
- changed `tsc-alias` with `tsconfig-paths` library (ONLY ON THE SAMPLE APP, I haven't touched the CLI code)
- brought back the `bootstrap` method in `index.ts`
- did a bit separation in the `src/common/` folder
- added versioning prefixes
- each route is now in a separate file
- added request schemas to the route definition
- services are in separate folder, not in `src/api/`

What I haven't addressed yet:
- auth strategies
- opt for using `passport` or `express-jwt`

Yavorss commented on assets/express/src/api/user/v1/index.ts:
```diff
@@ -0,0 +1,24 @@
+
+import { verifyJWT } from "@middleware";
+
+export const userControllerFactory: ControllerFactory<
```
I see this a little differently. The `v1` folder should just export all the routes and this ControllerFactory function should be generic for all the versions and modules.

Yavorss commented on assets/express/src/api/user/index.ts:
```diff
@@ -0,0 +1,14 @@
+import { ModuleFactory } from '@common';
+import { userServiceFactory } from '@services';
+import { UserController } from './interfaces';
+import { userControllerFactory } from './v1';
```
Here we can just have `import * as version1Controllers from './v1';` Otherwise for each version we would have
```
import { userControllerFactoryV1 } from './v1';
import { userControllerFactoryV2 } from './v2';
```

Yavorss commented on assets/express/src/api/user/index.ts:
```diff
@@ -0,0 +1,14 @@
+export const userModuleFactory: ModuleFactory<
+  UserController
+> = function ({ userRepoitory }) {
+  const userService = userServiceFactory(userRepoitory);
```
When we introduce DI, here instead of passing just the `userRepository` we need to provide all the possibilities. In real life, we would need different services and repositories basically everywhere. Sometimes you need to use an already implemented service method of another service or a repository, so you can implement the business logic.

Yavorss commented on assets/express/src/api/user/dto/user.dto.ts:
```diff
@@ -0,0 +1,38 @@
+import { z } from 'zod';
```
Having a file per route but one file for all DTOs is kinda weird to me... What about if version 2 of a route needs a change?

Yavorss commented on assets/express/src/api/user/dto/user.dto.ts:
```diff
@@ -0,0 +1,38 @@
+  }),
+});
+
+export const updateUserRequestPayload = User;
```
Instead of `RequestPayload` we can use `DTO` since `DTO` means Data Transfer Object or just nothing since this lives in the DTO folder.

Yavorss commented on assets/express/src/api/user/helpers/index.ts:
```diff
@@ -0,0 +1,31 @@
+  password: string,
+  hashedPassword: string
+): Promise<boolean> {
+  return bcrypt.compare(password, hashedPassword);
```
Pretty sure we already have these in the `services` folder.

Yavorss commented on assets/express/src/api/index.ts:
```diff
@@ -0,0 +1,21 @@
+  // todos: TodosController
+}
+
+export type ApiRoutes = UserController[keyof UserController] | HealthzController[keyof HealthzController]
```
I don't this would scale well and I don't see a reason of typing this. This is only used in the `app.js` but all we need is a generic route type that has `method, url, requestSchema, middlewares, handler` etc.

Yavorss commented on assets/express/src/services/user.service.ts:
```diff
@@ -0,0 +1,168 @@
+  ) {
+    const { email, password, ...attributes } = payload;
+    console.log({ payload });
+    const errors: Error[] = [];
```
Why do we have multiple errors?

Yavorss commented on assets/express/src/services/user.service.ts:
```diff
@@ -0,0 +1,168 @@
+
+    return {
+      errors,
+      status: !errors.length && user ? STATUS.SUCCESS: STATUS.FAILURE,
```
The status is something related to the route itself, not to the service. Services shouldn't know about routes in general, they just do the business logic.

Yavorss commented on assets/express/src/database/models/user.model.ts:
```diff
@@ -0,0 +1,45 @@
+// This is just sample implementation at this point, the DB functionality
+// will be added with next tickets
+// eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unused-vars
+export default function userOperations(dbConnection: any): UserRepository {
```
This is `user.model.ts` with function name `userOperations` that returns `UserRepository`?

Yavorss commented on assets/express/src/database/models/user.model.ts:
```diff
@@ -0,0 +1,45 @@
+    return dbConnection.users.get(email);
+  }
+
+  async function getAllUsers(): Promise<User[]> {
```
Since this is a repository, IMO we don't need to prefix or suffix it. Just getAll(), create(), delete(), get() is enough.

zhitomir-oreshenski-mm commented on assets/express/src/api/v1/todos/dto/replace-todo.dtm.ts:
```diff
@@ -0,0 +1,3 @@
+import { putTodoInput } from 'src/modules';
```
It seems that the extension is not right.

```suggestion
replace-todo.dto.ts instead replace-todo.dtm.ts
```

zhitomir-oreshenski-mm commented on assets/express/src/api/v1/todos/dto/update-todo.dtm.ts:
```diff
@@ -0,0 +1,3 @@
+import { patchTodoInput } from 'src/modules';
```
It seems that the extension is not right.

```suggestion
update-todo.dto.ts instead update-todo.dtm.ts
```

zhitomir-oreshenski-mm commented on assets/express/src/api/v1/todos/routes/delete-todo.ts:
```diff
@@ -0,0 +1,28 @@
+    description: 'Soft delete a To-Do item',
+    tags: ['Todo'],
+    method: 'delete',
+    path: '/todos/:id',
```
```suggestion
    path: '/:id',
```

zhitomir-oreshenski-mm replied:
Eg.

<img width="595" alt="image" src="https://user-images.githubusercontent.com/58042315/225098694-b3deaf03-95a3-41de-a57b-6cf850247d02.png">

zhitomir-oreshenski-mm commented on assets/express/src/api/v1/todos/routes/get-todo.ts:
```diff
@@ -0,0 +1,28 @@
+    description: 'Get a To-Do item',
+    tags: ['Todo'],
+    method: 'get',
+    path: '/todos/:id',
```
```suggestion
    path: '/:id',
```

zhitomir-oreshenski-mm commented on assets/express/src/api/v1/todos/routes/patch-todo.ts:
```diff
@@ -0,0 +1,30 @@
+    description: 'Partially update a To-Do item',
+    tags: ['Todo'],
+    method: 'patch',
+    path: '/todos/:id',
```
```suggestion
    path: '/:id',
```

zhitomir-oreshenski-mm commented on assets/express/src/api/v1/todos/routes/update-todo.ts:
```diff
@@ -0,0 +1,30 @@
+    description: 'Update a To-Do item',
+    tags: ['Todo'],
+    method: 'put',
+    path: '/todos/:id',
```
```suggestion
    path: '/:id',
```
